pandas/core/indexes/multi.py

```diff
@@ -1435,6 +1447,89 @@ def from_product(cls, iterables, sortorder=None, names=None):
 
+    @classmethod
+    def from_frame(cls, df, sortorder=None, names=None):
+        """
+        Make a MultiIndex from a DataFrame.
+
+        .. versionadded:: 0.24.0
+
+        Parameters
+        ----------
+        df : pd.DataFrame
+            DataFrame to be converted to MultiIndex.
+        sortorder : int or None
+            Level of sortedness (must be lexicographically sorted by that
+            level).
+        names : list-like, optonal
+            If no names are provided, use the column names, or tuple of column
+            names if the columns is a MultiIndex. If a sequence, overwrite
+            names with the given sequence.
+
+        Returns
+        -------
+        MultiIndex or Index
+            The MultiIndex representation of the given DataFrame.
+
+        Examples
+        --------
+        >>> df = pd.DataFrame([[0, 'happy'], [0, 'jolly'], [1, 'happy'],
+        ...                    [1, 'jolly'], [2, 'joy'], [2, 'joy']],
+        ...                   columns=['will_be', 'used'])
+        >>> df
+           will_be   used
+        0        0  happy
+        1        0  jolly
+        2        1  happy
+        3        1  jolly
+        4        2    joy
+        5        2    joy
+        >>> pd.MultiIndex.from_frame(df)
+        MultiIndex(levels=[[0, 1, 2], ['happy', 'jolly', 'joy']],
+                   labels=[[0, 0, 1, 1, 2, 2], [0, 1, 0, 1, 2, 2]],
+                   names=['will_be', 'used'])
+
+        >>> df = pd.DataFrame([['ahc', 'iam'], ['ahc', 'wim'], ['boh', 'amg'],
+        ...                    ['boh', 'iam'], ['oil', 'wim'], ['oil', 'amg']],
+        ...                   columns=['will_be', 'overriden'])
+        >>> df
+           will_be   overriden
+        0      ahc         iam
+        1      ahc         wim
+        2      boh         amg
+        3      boh         iam
+        4      oil         wim
+        5      oil         amg
+        >>> pd.MultiIndex.from_frame(df, names=['sure', 'will'])
+        MultiIndex(levels=[['ahc', 'boh', 'oil'], ['amg', 'iam', 'wim']],
+                   labels=[[0, 0, 1, 1, 2, 2], [1, 2, 0, 1, 2, 0]],
+                   names=['sure', 'will'])
+
+        See Also
+        --------
+        MultiIndex.from_arrays : Convert list of arrays to MultiIndex.
+        MultiIndex.from_tuples : Convert list of tuples to MultiIndex.
+        MultiIndex.from_product : Make a MultiIndex from cartesian product
+                                  of iterables.
+        """
+        if not isinstance(df, ABCDataFrame):
+            raise TypeError("Input must be a DataFrame")
+
+        column_names, columns = lzip(*df.iteritems())
+
+        # Get MultiIndex names
+        if names is None:
+            names = column_names
+        elif is_list_like(names):
+            if len(names) != len(df.columns):
+                    raise ValueError("'names' should have same length as "
+                                     "number of columns in df.")
```

Comment:
I suggested in my [previous comment](https://github.com/pandas-dev/pandas/pull/23141#discussion_r234456258) that all checks on ``names`` are superfluous: they exactly repeat the checks done inside the constructor. Isn't
```
names = column_names if names is None else names
```
sufficient?!

Reply:
Yup, removed the redundant code. Thanks.